RequireThis: Incorrectly triggers for record fields in compact constructor

Check documentation: https://checkstyle.org/config_coding.html#RequireThis

```
/var/tmp $ javac --enable-preview --release 14 RequireThisRecord.java
Note: RequireThisRecord.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
```

`/var/tmp $ cat checkstyle-recordthis.xml`
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <property name="severity" value="${severity}" default="warning"/>

    <property name="fileExtensions" value="java, properties, xml"/>
    <!-- Excludes all 'module-info.java' files              -->
    <!-- See https://checkstyle.org/config_filefilters.html -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$"/>
    </module>

    <module name="TreeWalker">
        <module name="RequireThis">
            <property name="validateOnlyOverlapping" value="false"/>
        </module>
    </module>
</module>
```

`/var/tmp $ cat RequireThisRecord.java`
```java
import java.util.Map;

record RequireThisRecord(String name, Map<String, String> items) {

    public RequireThisRecord {
        if (name == null) {
            name = "<unknown>";
        }
        items = Map.copyOf(items);
    }

}
```

```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-8.36.2-all.jar -c checkstyle-recordthis.xml RequireThisRecord.java
Starting audit...
[WARN] RequireThisRecord.java:6:13: Reference to instance variable 'name' needs "this.". [RequireThis]
[WARN] RequireThisRecord.java:7:13: Reference to instance variable 'name' needs "this.". [RequireThis]
[WARN] RequireThisRecord.java:9:9: Reference to instance variable 'items' needs "this.". [RequireThis]
[WARN] RequireThisRecord.java:9:28: Reference to instance variable 'items' needs "this.". [RequireThis]
Audit done.

```
---------------

I'm back again :) been playing with records some more in Configurate, and came across this issue. The `RequireThis` check seems to treat accesses in the record constructor as acting on the backing fields, not the (implicit) constructor parameters, so it expects them to be qualified with `this.`. However, adding `this.` to the variables results in a compile-time error, so this check should not trigger an error.

From [the JLS section on records](https://docs.oracle.com/javase/specs/jls/se15/preview/specs/records-jls.html#jls-8.10), section 8.10.4:

> It is a compile-time error if an assignment occurs (16) to a field corresponding to a record component of the record class in the body of the compact constructor.
>
> All fields corresponding to the record components of the record class are implicitly initialized to the value of the corresponding formal parameter after the body of the compact constructor. These fields are implicitly initialized in the order that they are declared in the record component list.

I came across this as part of some work I'm doing locally to validate the API -- the full original class that caused the error is at https://gist.github.com/zml2008/5f6de6484eed5b9d8451c5e1625f8572

@zml2008 @pbludov  Sorry for the delay.

```
$cat RequireThisRecord.java
import java.util.Map;

record RequireThisRecord(String name, Map<String, String> items) {

    public RequireThisRecord {
        if (name == null) {
            this.name = "<unknown>";
        } else {
            this.name = name;
        }
        this.items = Map.copyOf(items);
    }

}
```

Compile with Java 14.0.1: Compile successful
Compile with Java 15.0.1:
```
RequireThisRecord.java:7: error: cannot assign a value to final variable name
            this.name = "<unknown>";
                ^
RequireThisRecord.java:9: error: cannot assign a value to final variable name
            this.name = name;
                ^
RequireThisRecord.java:11: error: cannot assign a value to final variable items
        this.items = Map.copyOf(items);
            ^
Note: RequireThisRecord.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
3 errors
```

2 java versions give different compiling results. 
So that means we need to have different logic for each version?

@romani @zml2008  Sorry for the delay. I found out 1 more bug:

```
$ cat RequireThisRecord.java
record RequireThisRecord(String name) {
    public void foo(String name) {
        name = name;
    }
}

$ cat config.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="RequireThis">
            <property name="validateOnlyOverlapping" value="false"/>
        </module>
    </module>
</module>

javac --source 15 --enable-preview RequireThisRecord.java // java 14 and 15 compiles fine

java -jar .\checkstyle-8.37-all.jar -c config.xml RequireThisRecord.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing RequireThisRecord.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:311)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:221)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:408)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:331)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:190)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:125)
Caused by: java.lang.NullPointerException: Cannot invoke "com.puppycrawl.tools.checkstyle.api.DetailAST.findFirstToken(int)" because "mods" is null
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck$ClassFrame.hasFinalField(RequireThisCheck.java:1513)
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck.canAssignValueToClassField(RequireThisCheck.java:903)
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck.getClassFrameWhereViolationIsFound(RequireThisCheck.java:682)
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck.getFieldWithoutThis(RequireThisCheck.java:507)
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck.processIdent(RequireThisCheck.java:460)
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck.visitToken(RequireThisCheck.java:394)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:340)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:451)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:278)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:151)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:87)
        at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:337)
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:298)
        ... 5 more
Checkstyle ends with 1 errors.
```
Should I fix both bugs in the same PR, or should I create a separate issue for the new bug?

The change in Java 15 causes another (small) error in our check too:
```
$ cat InputRequireThis.java
public record InputRequireThis(int x, int y) {
  InputRequireThis(int x) {
    this(x, 42);
    x = x; // wrong error message
  }
}

java -jar .\checkstyle-8.37-all.jar -c config.xml InputRequireThis.java
Starting audit...
[ERROR] E:\checkstyle-playground\InputRequireThis.java:4:5: Reference to instance variable 'x' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 1 errors.
```

Because 'x' is `final` instance variable, `this.x = x` is not possible. However, the error is at 4:5 (The first 'x'), while it should be at 4:9 (The second 'x').

It seems that RequireThis hasn't handled the case where an instance variable is final:
```
$ cat Test.java
public class Test {
  final int x = 0;
  public Test(int x) {
    x = x;
  }
}

java -jar .\checkstyle-8.37-all.jar -c config.xml Test.java
Starting audit...
[ERROR] E:\checkstyle-playground\Test.java:5:5: Reference to instance variable 'x' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 1 errors.
```

Similar to [my previous comment](https://github.com/checkstyle/checkstyle/issues/8873#issuecomment-723727095), error should occur at 5:9 since `this.x = x` gives a compile error (`x` is `final`).
I think this problem is more about `RequireThis` for `final variable` than about `compact constructor`. I'm gonna make a commit to fix `compact constructor` problem in this issue, and then open another issue for `final variable`.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java
@@ -458,7 +458,8 @@ public class RequireThisCheck extends AbstractCheck {
             default:
                 if (checkFields) {
                     final AbstractFrame frame = getFieldWithoutThis(ast, parentType);
-                    if (frame != null) {
+                    final boolean canUseThis = !isInCompactConstructor(ast);
+                    if (frame != null && canUseThis) {
                         logViolation(MSG_VARIABLE, ast, frame);
                     }
                 }
@@ -508,6 +509,25 @@ public class RequireThisCheck extends AbstractCheck {
             }
         }
         return frame;
+    }
+
+    /**
+     * Return whether ast is in a COMPACT_CTOR_DEF.
+     *
+     * @param ast The token to check
+     * @return true if ast is in a COMPACT_CTOR_DEF, false otherwise
+     */
+    private static boolean isInCompactConstructor(DetailAST ast) {
+        boolean isInCompactCtor = false;
+        DetailAST parent = ast.getParent();
+        while (parent != null) {
+            if (parent.getType() == TokenTypes.COMPACT_CTOR_DEF) {
+                isInCompactCtor = true;
+                break;
+            }
+            parent = parent.getParent();
+        }
+        return isInCompactCtor;
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheckTest.java
@@ -424,6 +424,16 @@ public class RequireThisCheckTest extends AbstractModuleTestSupport {
     }
 
     @Test
+    public void testRecordCompactCtors() throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(RequireThisCheck.class);
+        checkConfig.addAttribute("validateOnlyOverlapping", "false");
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verify(checkConfig,
+                getNonCompilablePath("InputRequireThisRecordCompactCtors.java"),
+                expected);
+    }
+
+    @Test
     public void testRecordsAsTopLevel() throws Exception {
         final DefaultConfiguration checkConfig = createModuleConfig(RequireThisCheck.class);
         checkConfig.addAttribute("checkFields", "true");
@@ -432,10 +442,6 @@ public class RequireThisCheckTest extends AbstractModuleTestSupport {
             "14:9: " + getCheckMessage(MSG_METHOD, "method1", ""),
             "15:9: " + getCheckMessage(MSG_METHOD, "method2", ""),
             "16:9: " + getCheckMessage(MSG_METHOD, "method3", ""),
-            "17:17: " + getCheckMessage(MSG_VARIABLE, "x", ""),
-            "17:21: " + getCheckMessage(MSG_VARIABLE, "y", ""),
-            "18:28: " + getCheckMessage(MSG_VARIABLE, "y", ""),
-            "18:32: " + getCheckMessage(MSG_VARIABLE, "x", ""),
             "23:9: " + getCheckMessage(MSG_METHOD, "method1", ""),
             "27:21: " + getCheckMessage(MSG_VARIABLE, "x", ""),
             "35:17: " + getCheckMessage(MSG_VARIABLE, "y", ""),

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/requirethis/InputRequireThisRecordAsTopLevel.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/requirethis/InputRequireThisRecordAsTopLevel.java
@@ -14,8 +14,8 @@ public record InputRequireThisRecordAsTopLevel(int x, int y) {
         method1(); // violation
         method2(42); // violation
         method3(); // violation
-        int z = x + y + 2; // violation x2
-        System.out.println(y + x); // violation x2
+        int z = x + y + 2; // ok, 'this' cannot be used in COMPACT_CTOR_DEF
+        System.out.println(y + x); // ok, 'this' cannot be used in COMPACT_CTOR_DEF
     }
 
     InputRequireThisRecordAsTopLevel(int x) {

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/requirethis/InputRequireThisRecordCompactCtors.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/requirethis/InputRequireThisRecordCompactCtors.java
@@ -1,0 +1,17 @@
+//non-compiled with javac: Compilable with Java14
+package com.puppycrawl.tools.checkstyle.checks.coding.requirethis;
+import java.util.Map;
+
+/* Config:
+ * validateOnlyOverlapping = false
+ *
+ */
+public record InputRequireThisRecordCompactCtors(String name, Map<String, String> items){
+    // 'this' cannot be used in compact constructor
+    public InputRequireThisRecordCompactCtors {
+        if (name == null) { // ok
+            name = "<unknown>"; // ok
+        }
+        items = Map.copyOf(items); // ok
+    }
+}

